gitlab-runner/src/run.rs

```diff
@@ -61,7 +63,33 @@ where
         }
     } else {
         script_result
-    };
+    }
+}
+
+async fn run<F, J, Ret>(
+    job: Job,
+    client: Client,
+    response: Arc<JobResponse>,
+    process: F,
+    build_dir: PathBuf,
+    cancel_token: CancellationToken,
+) -> JobResult
+where
+    F: FnOnce(Job) -> Ret,
+    J: JobHandler,
+    Ret: Future<Output = Result<J, ()>>,
+{
+    if let Err(e) = tokio::fs::create_dir(&build_dir).await {
+        job.trace(format!("Failed to remove build dir: {}", e));
+        return Err(());
+    }
+    let mut handler = cancellable_wait(process(job), &cancel_token).await?;
```

Comment:
So process is meant to be the crates users factory function (iotw it crates the handler). As part of that it might do some setup including local setup.  So i don't think it should be cancellable by gitlab.

It seems better to have this not be cancellable and then to have the job immediately cancel when the running of the steps happens at which point cleanup also gets invoked

Fwiw all this behaviour should be documented so the crate user knows what to expect; Especially when doing cleanup so they know what might be needed to cleanup..  As a side-note  it might be worth considering adding a explicit hint the cleanup is called due to a cancellation?